File: Ridge Regression (L2 regularization)/Ridge_regression.py
import numpy as np
x = np.array([1,2,3,4,5,6,7,8,9,10],dtype=float)
y = np.array([1,3,5,7,9,11,13,15,17,19],dtype=float)
w = 0.0
b = 0.0
lr = 0.01
epoch = 10000
lambda_ = 0.1
for i in range(epoch):
    pred = w * x + b
    loss = np.mean((pred - y)**2) + lambda_ * w**2

    dw = (-2 / len(x)) * np.sum(x * (y - pred)) + 2 * lambda_ * w
    db = (-2 / len(x)) * np.sum(y - pred)

    w -= lr * dw
    b -= lr * db

    if i % 1000 == 0:
        print(f"epoch: {i} | loss: {loss:.6f} | w: {w:.4f} | b: {b:.4f}")

print(f"\nPrediction for x=7: {w*7+b:.4f} (expected 14)")

File: Ridge Regression (L2 regularization)/Ridge_regression.c
#include <stdio.h>
#define N 10
#define LR 0.01
#define EPOCH 5000
#define LAMBDA 0.01
int main()
{
    double x[] = {1,2,3,4,5,6,7,8,9,10};
    double y[] = {1,3,5,7,9,11,13,15,17,19};
    double w = 0;
    double b = 0;
    for (int epoch = 0; epoch < EPOCH; epoch++)
    {
        double loss = 0;
        for (int i = 0; i < N; i++)
        {
            double pred = w*x[i] + b;
            loss += (y[i] - pred)*(y[i] - pred);
        }
        loss /= N;
        loss += LAMBDA * w * w;
        if (loss < 0.000001)
        {
            break;
        }
        double dw = 0;
        double db = 0;
        for (int i = 0; i < N; i++)
        {
            double pred = w*x[i] + b;
            dw += -2.0/N * x[i] * (y[i] - pred);
            db += -2.0/N * (y[i] - pred);
        }
        dw += 2 * LAMBDA * w;
        w -= LR * dw;
        b -= LR * db;
        if (epoch % 1000 == 0)
        {
            printf("epoch: %d, w: %f, b: %f, loss: %f\n", epoch, w, b, loss);
        }
    }
    printf("\nFinal: w = %f, b = %f\n", w, b);
    printf("\nPredictions:\n");
    for (int i = 0; i < N; i++)
    {
        double pred = w * x[i] + b;
        printf("x: %.0f | pred: %.2f | actual: %.0f\n", x[i], pred, y[i]);
    }
}